### PR TITLE
[FLINK-7452] [types] Add helper methods for all built-in Flink types

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This PR provides an unified, detailed, and easy to use utility to work with Flink's built-in Java types.

This class helps users to get an overview about built-in Flink types and their features/limitations (esp. regarding null support).

## Brief change log

Improved `org.apache.flink.api.common.typeinfo.Types`.

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

**Comment by twalthr** [comment]
CC @alpinegizmo

**Comment by twalthr** [comment]
@fhueske Feel free to review.

File: flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java [hunk 1]
```diff
@@ -19,56 +19,408 @@
...
+	 * about how to improve efficiency (namely by pre-registering classes).
+	 *
+	 * @param genericClass any Java class
+	 */
+	public static TypeInformation<?> GENERIC(Class<?> genericClass) {
+		return new GenericTypeInfo<>(genericClass);
+	}
+
+	/**
+	 * Returns type information for Java arrays of primitive type (such as <code>byte[]</code>). The array
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java]: `The array and its elements do not support null values.` -> `The array must not be null.`? Entries of primitive arrays cannot be null anyway.

File: flink-examples/flink-examples-batch/src/main/scala/org/apache/flink/examples/scala/wordcount/WordCount.scala [hunk 2]
```diff
@@ -49,6 +51,8 @@ object WordCount {
 
     val params: ParameterTool = ParameterTool.fromArgs(args)
 
+    val t = Types.of[Either[String, Nothing]]
```
Comment (fhueske) [inline flink-examples/flink-examples-batch/src/main/scala/org/apache/flink/examples/scala/wordcount/WordCount.scala]: What's the purpose of this change?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/Types.scala [hunk 3]
```diff
@@ -20,60 +20,123 @@ package org.apache.flink.table.api
...
 import org.apache.flink.table.typeutils.TimeIntervalTypeInfo
-import org.apache.flink.types.Row
 
 import _root_.scala.annotation.varargs
 
 /**
-  * This class enumerates all supported types of the Table API.
+  * This class enumerates all supported types of the Table API & SQL.
   */
 object Types {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/Types.scala]: Not sure if that's possible, but could the `Types` object extend the `JTypes` class such that we don't have to implement all methods again and forward the calls?

File: flink-scala/src/main/scala/org/apache/flink/api/scala/Types.scala [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+  * (or inefficient) as well as cases where type information has to be supplied manually.
+  *
+  * Depending on the API you are using (e.g. Java API or Table API), there might be a more
+  * specialized `Types` class.
+  *
+  * Scala macros allow to determine type information of classes and type parameters. You can
+  * use [[Types.of]] to let type information be determined automatically.
+  */
+@PublicEvolving
+object Types {
```
Comment (fhueske) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/Types.scala]: extend the Java `Types` class?

File: flink-scala/src/main/scala/org/apache/flink/api/scala/Types.scala [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+  * has built-in serializers and comparators.
+  *
+  * This class contains types of [[org.apache.flink.api.common.typeinfo.Types]] and adds
+  * types for Scala specific classes (such as [[Unit]] or case classes).
+  *
+  * In many cases, Flink tries to analyze generic signatures of functions to determine return
+  * types automatically. This class is intended for cases where the extraction is not possible
+  * (or inefficient) as well as cases where type information has to be supplied manually.
+  *
+  * Depending on the API you are using (e.g. Java API or Table API), there might be a more
```
Comment (fhueske) [inline flink-scala/src/main/scala/org/apache/flink/api/scala/Types.scala]: Remove this comment?

File: flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java [hunk 6]
```diff
@@ -19,56 +19,408 @@
...
+
+	/**
+	 * Returns type information for a POJO (Plain Old Java Object).
+	 *
+	 * <p>A POJO class is public and standalone (no non-static inner class). It has a public no-argument
+	 * constructor. All non-static, non-transient fields in the class (and all superclasses) are either public
+	 * (and non-final) or have a public getter and a setter method that follows the Java beans naming
+	 * conventions for getters and setters.
+	 *
+	 * <p>A POJO is a fixed-length, null-aware composite type with non-deterministic field order. Every field
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java]: Table API or a `keyBy(0, 1)` operation.

File: flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java (hunk 1 above)
Comment (twalthr) [reply flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java]: I just wanted to be consistent :D

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/Types.scala (hunk 3 above)
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/Types.scala]: We had this initially. The problem is that Scala creates method where as Java fields. It would be a mixture of `STRING()`, `INTERVAL_MILLIS`. Furthermore, not all types of the Java API are important for Table API users. I think it is better to have specialized classes for each API. This also allows to have custom Javadocs. Like the STRING <-> VARCHAR mapping comment.

File: flink-examples/flink-examples-batch/src/main/scala/org/apache/flink/examples/scala/wordcount/WordCount.scala (hunk 2 above)
Comment (twalthr) [reply flink-examples/flink-examples-batch/src/main/scala/org/apache/flink/examples/scala/wordcount/WordCount.scala]: Good question. Will remove it.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/Types.scala (hunk 3 above)
Comment (twalthr) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/Types.scala]: If you compare the two classes, you will notice that they are not the same. There are differences.

**Comment by twalthr** [comment]
Thanks for the review @fhueske. I hope I addressed all your comments. I would not let the other classes extend the Java class. There are multiple differences. E.g. `Either` and `Option` in Scala/Java.

File: flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (fhueske) [reply flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java]: Key selection by index is not supported for POJOs and yields an exception. 
The case of importing a POJO DataStream as Table is described in the Table API / SQL documentation.

I would not add this detail here. IMO, it rather confuses users than helps them.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/Types.scala (hunk 3 above)
Comment (fhueske) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/Types.scala]: Makes sense, thanks

File: flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java [hunk 7]
```diff
@@ -19,56 +19,417 @@
...
+import java.util.Map;
 
 /**
- * This class gives access to the type information of the most most common types.
+ * This class gives access to the type information of the most common types for which Flink
+ * has built-in serializers and comparators.
+ *
+ * <p>In many cases, Flink tries to analyze generic signatures of functions to determine return
+ * types automatically. This class is intended for cases where type information has to be
+ * supplied manually or would result in an inefficient type.
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java]: "or would result in an inefficient type" -> "or cases where automatic type inference results in an inefficient type."

File: flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java [hunk 8]
```diff
@@ -19,56 +19,417 @@
...
+	public static final TypeInformation<?> SQL_TIMESTAMP = SqlTimeTypeInfo.TIMESTAMP;
 
 	/**
-	 * Generates a RowTypeInfo with fields of the given types.
-	 * The fields have the default names (f0, f1, f2 ..).
+	 * Returns type information for {@link org.apache.flink.types.Row} with fields of the given types.
+	 * A row itself must not be null.
+	 *
+	 * <p>A row is a fixed-length, null-aware composite type for storing multiple values in a
+	 * deterministic field order. Every field can be null independent of the field's type.
```
Comment (fhueske) [inline flink-core/src/main/java/org/apache/flink/api/common/typeinfo/Types.java]: "independent of" -> "regardless of"?

**Comment by twalthr** [comment]
@fhueske I updated this PR and added generic parameters as well.

**Comment by twalthr** [comment]
@fhueske I would merge this later today if there are no objections.
